src/main/java/fi/pohina/vinkkilista/domain/TagService.java

```diff
@@ -0,0 +1,128 @@
+package fi.pohina.vinkkilista.domain;
+
+import fi.pohina.vinkkilista.data_access.TagDao;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.UUID;
+
+public class TagService {
+
+    private final TagDao tagDao;
+
+    public TagService(TagDao tagDao) {
+        this.tagDao = tagDao;
+    }
+
+    /**
+     * Finds existing {@link Tag} using the specified name, otherwise creates a
+     * new one.
+     *
+     * @param name the name of the new tag to be created
+     * @return reference to the matching or new tag
+     */
+    public Tag findOrCreateTag(String name) {
+        name = validateTag(name);
+        if (name.isEmpty()) {
+            return null;
+        }
+
+        Tag tag = tagDao.findByName(name);
+        if (tag != null) {
+            return tag;
+        }
+
+        String id = generateTagId();
+
+        tag = new Tag(id, name);
+
+        tagDao.add(tag);
+
+        return tag;
+    }
+
+    /**
+     * Function for converting string tag set to a set of tag objects, asking
+     * for new tags to be created if no matching ones are found. Validates the
+     * given set of tags.
+     *
+     * @param tagNames set of tag strings
+     * @return set of tag objects
+     */
+    public Set<Tag> findOrCreateTags(Set<String> tagNames) {
+        Set<Tag> tagsSet = new HashSet<>();
+
+        for (String name : tagNames) {
+            Tag tag = findOrCreateTag(name);
+            if (tag != null) {
+                tagsSet.add(tag);
+            }
+        }
+
+        return tagsSet;
+    }
+    
+    /**
+     * Function for getting a tag name by url.
+     *
+     * @param url the url to be analyzed
+     * @return name of tag or null if none was found
+     */
+    public String tagFromUrl(String url) {
```

Comment:
I think we should refactor this method, something like this:
- have a `private final HashMap<String,String> urlToTagMap` that maps urls to matching tags
- initiate the map's contents in helper methods, e.g. `private void addUrlToVideoToMap()` adding respective `<urlPart,tagName>` String pairs to the map. These initiator methods could be called in the constructor.
- simplify `tagFromUrl` to single for loop going through each key `urlPart` in the map `keySet()`, and upon match `url.contains(urlPart)` returns `urlToTagMap.get(urlPart)`

Reply:
Sounds good, should be done like this!

Reply:
Another thing I realised in this class is that `validateTag(String tag)` functionality is already provided by `TagParser` class except for removal of empty strings post-trimming. As such it feels like `tagFromUrl()` could as well be in `TagParser` or its separate class so that `TagService` is only concerned with handling `Tag` objects.